### ISSUE TITLE
UCX Linux CI frequently times out

### ISSUE THREAD

**Issue by evan-charmworks** [UIUC-PPL/charm#2732]
The GitHub CI build for UCX Linux regularly fails due to taking more time than the 60 minute timeout we have specified, which is already longer than the 45 minute timeout for other builds. When the timeout occurs, `make test-tests test-examples` has been running for 35 minutes, but when the build completes successfully, the test step only takes 11 minutes.

**Comment by matthiasdiener** [comment]
Seems to hang in the `examples/charm++/load_balancing/stencil3d` test (see the timestamps on the left):

```
../../../../bin/testrun  +p4 ./stencil3d 64 32 +balancer DistributedLB +LBDebug 1  
2020-02-27T17:43:28.3268399Z 
2020-02-27T17:43:28.3269138Z Running on 4 processors:  ./stencil3d 64 32 +balancer DistributedLB +LBDebug 1 
2020-02-27T17:43:28.3326320Z charmrun>  /usr/bin/setarch x86_64 -R  mpirun -np 4  ./stencil3d 64 32 +balancer DistributedLB +LBDebug 1 
2020-02-27T17:43:28.3919884Z Charm++> Running in non-SMP mode: 4 processes (PEs)
2020-02-27T17:43:28.3920500Z Converse/Charm++ Commit ID: v6.10.0-rc2-158-g5cdb2cb97
2020-02-27T17:43:28.4054224Z Isomalloc> Synchronized global address space.
2020-02-27T17:43:28.4060522Z CharmLB> Verbose level 1, load balancing period: 0.5 seconds
2020-02-27T17:43:28.4060827Z CharmLB> Load balancer assumes all CPUs are same.
2020-02-27T17:43:28.4174003Z Charm++> Running on 1 hosts (1 sockets x 2 cores x 1 PUs = 2-way SMP)
2020-02-27T17:43:28.4613310Z Charm++> cpu topology info is gathered in 0.054 seconds.
2020-02-27T17:43:28.4692479Z CharmLB> DistributedLB created: threshold 0.000000, max phases 0
2020-02-27T17:43:28.4692540Z 
2020-02-27T17:43:28.4692789Z STENCIL COMPUTATION WITH BARRIERS
2020-02-27T17:43:28.4692908Z Running Stencil on 4 processors with (2, 2, 2) chares
2020-02-27T17:43:28.4692996Z Array Dimensions: 64 64 64
2020-02-27T17:43:28.4693148Z Block Dimensions: 32 32 32
2020-02-27T17:43:28.4823061Z [1] Time per iteration: 0.020298 0.088817
2020-02-27T17:43:28.5501299Z [2] Time per iteration: 0.067827 0.156659
2020-02-27T17:43:28.6222393Z [3] Time per iteration: 0.072074 0.228748
2020-02-27T17:43:28.6701159Z [4] Time per iteration: 0.047884 0.276664
2020-02-27T17:43:28.6981328Z [5] Time per iteration: 0.027969 0.304649
2020-02-27T17:43:28.9814327Z [DistributedLB] Load balancing step 0 starting at 0.587864
2020-02-27T17:43:28.9814918Z DistLB> DistributedLB: step 0 starting at 0.587893 Memory: 4.271606 MB
2020-02-27T17:43:28.9815062Z In DistributedLB strategy at 0.587903
2020-02-27T17:43:28.9972944Z DistributedLB>>>Before LB: max = 0.140701, avg = 0.077684, ratio = 1.811195
2020-02-27T17:43:29.0613947Z DistributedLB>>>Gossip finished at 0.667858 (0.079955 elapsed)
2020-02-27T17:43:29.0933315Z DistributedLB>>>After phase 1: max = 0.119819, avg = 0.077684, ratio = 1.542395
2020-02-27T17:43:29.1067515Z DistributedLB>>>After phase 2: max = 0.119819, avg = 0.077684, ratio = 1.542395
2020-02-27T17:43:29.1212999Z DistributedLB>>>After phase 3: max = 0.119819, avg = 0.077684, ratio = 1.542395
2020-02-27T17:43:29.1372804Z DistributedLB>>>After phase 4: max = 0.119819, avg = 0.077684, ratio = 1.542395
2020-02-27T17:43:29.1498173Z DistributedLB>>>After phase 5: max = 0.119819, avg = 0.077684, ratio = 1.542395
2020-02-27T17:43:29.1498403Z DistributedLB>>>Balancing completed at 0.756436 (0.168533 elapsed)
2020-02-27T17:43:29.1500218Z DistributedLB>>>1 total migrations with 1 negative ack max
2020-02-27T17:43:29.1500870Z DistributedLB> Strategy took 0.168752s memory usage: 4.274155 MB.
2020-02-27T17:43:29.1502532Z DistributedLB> step 0 finished at 0.756850 duration 0.168957 memory usage: 4.274063
[...]
2020-02-27T17:43:32.4145495Z [61] Time per iteration: 0.215872 3.860869
2020-02-27T17:43:32.4145735Z [62] Time per iteration: 0.015410 3.883340
2020-02-27T17:43:32.4145934Z [63] Time per iteration: 0.033586 3.916966
2020-02-27T17:43:32.4146168Z [64] Time per iteration: 0.035963 3.952981
2020-02-27T17:43:32.4146366Z [65] Time per iteration: 0.019900 3.972961
2020-02-27T17:43:32.4146568Z [DistributedLB] Load balancing step 12 starting at 3.991595
2020-02-27T17:43:32.4146816Z DistLB> DistributedLB: step 12 starting at 4.000135 Memory: 8.363281 MB
2020-02-27T17:43:32.4147026Z In DistributedLB strategy at 4.000177
2020-02-27T17:43:32.4147271Z DistributedLB>>>Before LB: max = 0.097203, avg = 0.050584, ratio = 1.921596
2020-02-27T17:43:32.4533521Z DistributedLB>>>Gossip finished at 4.059877 (0.059699 elapsed)
2020-02-27T17:43:32.4853305Z DistributedLB>>>After phase 1: max = 0.071204, avg = 0.050584, ratio = 1.407629
2020-02-27T17:43:32.5114546Z DistributedLB>>>After phase 2: max = 0.071204, avg = 0.050584, ratio = 1.407629
2020-02-27T17:43:32.5454258Z DistributedLB>>>After phase 3: max = 0.071204, avg = 0.050584, ratio = 1.407629
2020-02-27T17:43:32.5612954Z DistributedLB>>>After phase 4: max = 0.071204, avg = 0.050584, ratio = 1.407629
2020-02-27T17:43:32.5812930Z DistributedLB>>>After phase 5: max = 0.071204, avg = 0.050584, ratio = 1.407629
2020-02-27T17:43:32.5814385Z DistributedLB>>>Balancing completed at 4.188052 (0.187875 elapsed)
2020-02-27T17:43:32.5815768Z DistributedLB>>>3 total migrations with 1 negative ack max
2020-02-27T17:43:32.5817195Z DistributedLB> Strategy took 0.188198s memory usage: 8.363281 MB.
2020-02-27T17:43:32.5926004Z DistributedLB> step 12 finished at 4.199172 duration 0.199037 memory usage: 8.363281
2020-02-27T18:13:40.1113931Z ##[error]The operation was canceled.

**Comment by matthiasdiener** [comment]
Also saw a very similar error with netlrts (https://circleci.com/gh/UIUC-PPL/charm/4358?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link):

```
../../../../bin/testrun  +p4 ./stencil3d 64 32 +balancer DistributedLB +LBDebug 1  ++local
Charmrun> scalable start enabled. 
Charmrun> started all node programs in 0.008 seconds.
Charm++> Running in non-SMP mode: 4 processes (PEs)
Converse/Charm++ Commit ID: v6.10.0-rc2-159-g8097f6da4
Isomalloc> Synchronized global address space.
Charm++> scheduler running in netpoll mode.
CharmLB> Verbose level 1, load balancing period: 0.5 seconds
CharmLB> Load balancer assumes all CPUs are same.
Charm++> Running on 1 hosts (1 sockets x 1 cores x 2 PUs = 2-way SMP)
Charm++> cpu topology info is gathered in 0.007 seconds.
CharmLB> DistributedLB created: threshold 0.000000, max phases 0

STENCIL COMPUTATION WITH BARRIERS
Running Stencil on 4 processors with (2, 2, 2) chares
Array Dimensions: 64 64 64
Block Dimensions: 32 32 32
[1] Time per iteration: 0.106664 0.141415
[2] Time per iteration: 0.274145 0.415663
[3] Time per iteration: 0.232223 0.657341
[4] Time per iteration: 0.259564 0.917198
[5] Time per iteration: 0.255733 1.173188
[DistributedLB] Load balancing step 0 starting at 1.397234
DistLB> DistributedLB: step 0 starting at 1.397485 Memory: 1.676865 MB
In DistributedLB strategy at 1.397589
DistributedLB>>>Before LB: max = 1.230591, avg = 0.852778, ratio = 1.443038
DistributedLB>>>Gossip finished at 1.473232 (0.075643 elapsed)
DistributedLB>>>After phase 1: max = 1.230591, avg = 0.852778, ratio = 1.443038
DistributedLB>>>After phase 2: max = 1.230591, avg = 0.852778, ratio = 1.443038
DistributedLB>>>After phase 3: max = 1.230591, avg = 0.852778, ratio = 1.443038
DistributedLB>>>After phase 4: max = 1.230591, avg = 0.852778, ratio = 1.443038
DistributedLB>>>Balancing completed at 1.533469 (0.135880 elapsed)
DistributedLB>>>0 total migrations with 0 negative ack max
DistributedLB> Strategy took 0.136185s memory usage: 1.678772 MB.
DistributedLB> step 0 finished at 1.533806 duration 0.136321 memory usage: 1.710602
[6] Time per iteration: 0.382677 1.556112
[7] Time per iteration: 0.248901 1.814021
[8] Time per iteration: 0.254366 2.068676
[9] Time per iteration: 0.251941 2.320925
[10] Time per iteration: 0.239429 2.568551
[DistributedLB] Load balancing step 1 starting at 2.783734
DistLB> DistributedLB: step 1 starting at 2.783954 Memory: 1.680176 MB
In DistributedLB strategy at 2.784023
DistributedLB>>>Before LB: max = 1.206479, avg = 0.852510, ratio = 1.415208
DistributedLB>>>Gossip finished at 2.869169 (0.085145 elapsed)
DistributedLB>>>After phase 1: max = 1.206479, avg = 0.852510, ratio = 1.415208
DistributedLB>>>After phase 2: max = 0.936869, avg = 0.852510, ratio = 1.098954
DistributedLB>>>After phase 3: max = 0.878051, avg = 0.852510, ratio = 1.029959
DistributedLB>>>Balancing completed at 2.937438 (0.153414 elapsed)
DistributedLB>>>2 total migrations with 1 negative ack max
DistributedLB> Strategy took 0.153676s memory usage: 1.681046 MB.
DistributedLB> step 1 finished at 3.145914 duration 0.361960 memory usage: 2.408478
[11] Time per iteration: 0.726677 3.295552
make[4]: *** wait: No child processes.  Stop.
make[4]: *** Waiting for unfinished jobs....
make[4]: *** wait: No child processes.  Stop.
make[3]: *** wait: No child processes.  Stop.
make[3]: *** Waiting for unfinished jobs....
make[3]: *** wait: No child processes.  Stop.
Makefile:50: recipe for target 'test-load_balancing' failed
make[2]: *** [test-load_balancing] Error 2
Makefile:28: recipe for target 'test-charm++' failed
make[1]: *** [test-charm++] Hangup
Makefile:188: recipe for target 'test-examples' failed
make: *** [test-examples] Hangup
```

**Comment by evan-charmworks** [comment]
Could be related to #2630, #2638, #2639

**Comment by epmikida** [comment]
The hang is weird in that it looks like it happens after the whole strategy is run to completion. I'll try to reproduce and diagnose it on my own machine this week.

**Comment by matthiasdiener** [comment]
#2629 might also be related.

**Comment by epmikida** [comment]
Can't seem to reproduce the netlrts hang. Tried it both on darwin and on linux, on current master, and on 6.10.1.

**Comment by epmikida** [comment]
To reproduce a related crash on netlrts, @kavithachandrasekar found this command:

./charmrun +p8 ++local ./stencil3d 128 32 +balancer DistributedLB +LBDebug 1 +pemap 0-1

Works on netlrts-linux smp builds without any other modifications. The crash is due to a PE attempting to migrate an object to itself. It doesn't seem to reproduce the hang for me, but the crash might be masking it as it happens in almost every run.

**Comment by matthiasdiener** [comment]
Fixed by upgrading to UCX 1.9.0.
